plugin/path_checkouts.go

```diff
@@ -5,11 +5,325 @@ import (
 	"fmt"
 	"time"
 
+	"github.com/hashicorp/go-multierror"
 	"github.com/hashicorp/vault/sdk/framework"
 	"github.com/hashicorp/vault/sdk/logical"
 )
 
-func (b *backend) pathReserveStatus() *framework.Path {
+const secretAccessKeyType = "creds"
+
+// TODO we're going to need to deal with mutexes.
+func (b *backend) pathSetCheckOut() *framework.Path {
+	return &framework.Path{
+		Pattern: libraryPrefix + framework.GenericNameRegex("name") + "/check-out$",
+		Fields: map[string]*framework.FieldSchema{
+			"name": {
+				Type:        framework.TypeLowerCaseString,
+				Description: "Name of the set",
+				Required:    true,
+			},
+			"ttl": {
+				Type:        framework.TypeDurationSecond,
+				Description: "In seconds, the length of time before the check-out will expire.",
+			},
+		},
+		Operations: map[logical.Operation]framework.OperationHandler{
+			logical.ReadOperation: &framework.PathOperation{
+				Callback: b.operationSetCheckOut,
+				Summary:  "Check a service account out from the library.",
+			},
+		},
+		HelpSynopsis: `Check a service account out from the library.`,
+	}
+}
+
+func (b *backend) operationSetCheckOut(ctx context.Context, req *logical.Request, fieldData *framework.FieldData) (*logical.Response, error) {
+	setName := fieldData.Get("name").(string)
+
+	ttlPeriodRaw, ttlPeriodSent := fieldData.GetOk("ttl")
+	if !ttlPeriodSent {
+		ttlPeriodRaw = 0
+	}
+	requestedTTL := time.Duration(ttlPeriodRaw.(int)) * time.Second
+
+	set, err := readSet(ctx, req.Storage, setName)
+	if err != nil {
+		return nil, err
+	}
+	if set == nil {
+		return logical.ErrorResponse(`"%s" doesn't exist`, setName), nil
+	}
+
+	// Find the first service account available.
```

Comment:
What is the plan for contention here? Will we be locking the backend, or retrying if an available accounts turns out to be unavailable during the actual checkout?

I saw in a previous resolved comment that locking was out of scope in this PR. I'd personally prefer that it be included, primarily for efficiency. It takes me some time to build a mental model of what an operation/function is doing, and then review it for correctness. When reviewing the locks/mutex PR, I'll have to re-familiarize myself with the same code again.

Reply:
Ok, I can add locking to this PR. I haven't fully thought through the locking model and I thought it would be most apparent after everything else had settled. I do think the code's together enough now to address it. I will probably end up pushing code around locking by the end of the week and I'll re-request a review at that time to indicate it's ready for a look.

Reply:
This is pushed up with tests.